rfcs/20250910-graph-api-dropout/README.md

```diff
@@ -0,0 +1,183 @@
+> Y_i = \frac{X_i \cdot M_i}{1-p}
+> $$
+
+where $M$ is the mask tensor (same shape as $X$), and $Y$ is the output tensor.
+
+**Forward (Inference)**
+
+Dropout is typically bypassed during inference:
+
+> $$
+> Y_i = X_i
+> $$
+
+**Backward**
+
+Given gradient tensor $dY$, mask tensor $M$, and dropout probability $p$:
+
+> $$
+> dX_i = \frac{dY_i \cdot M_i}{1-p}
+> $$
+
+where $dX$ is the gradient with respect to the input.
+
+## Adoption in Frameworks and Libraries
+
+**PyTorch**
+
+troch.nn.Dropout [[#2]][2] takes `p` (probability) and `inplace` as attributes,
+with one input and one output. Internally, the random number generator used to
+generate the dropout mask differs by device:
+
+- CPU: Uses torch.Generator backed by the C++ Mersenne Twister engine [[#3]][3],
+  which consumes a 64-bit seed.
+- CUDA / XPU: Uses a 64-bit Philox RNG state (seed + offset) obtained from the
+  device generator [[#4]][4][[#5]][5]. The offset is advanced after each kernel
+  launch to carve out disjoint subsequences of random numbers.
+
+**TensorFlow**
+
+tf.keras.layers.Dropout [[#6]][6] takes `rate` (probability), an optional `noise_shape`
+(shape for randomly generated keep/drop flags) and an optional seed as
+attributes, with one input and one output. Internally, the random number
+generator produces a 64-bit pair (key, counter) from the seed and passes
+them to a selected RNG algorithm to generate the mask.
+Supported algorithms [[#7]][7] include Philox, ThreeFry, or AUTO_SELECT
+(which lets TensorFlow choose the best algorithm for the current device;
+AUTO_SELECT may pick an algorithm other than Philox or ThreeFry).
+
+**ONNX**
+
+ONNX's Dropout operation [[#8]][8] takes `data`, `ratio` (probability) and `training_mode`
+as inputs, and `seed` as an attribute. Outputs include the dropped tensor and
+the dropout mask. ONNX does not specify the algorithm used to generate random numbers.
+
+**cuDNN**
+
+cuDNN Graph library does not provide a dedicated Dropout operation. Instead, it
+introduces an RNG operation, which can be used to implement Dropout with two
+pointwise Multiply operations.
+The RNG operation takes `seed`, `offset` as inputs, supports various
+distributions (Bernoulli, Uniform, Normal) by setting attributes, and outputs
+the generated tensor. The RNG is based on the Philox algorithm.
+
+**oneDNN**
+
+In oneDNN primitive, Dropout is implemented as a binary post-op applied to
+output tensor before other post-ops. It uses the Philox algorithm and takes
+`seed` (s32) and `probability` (f32) as inputs, and outputs a `mask` (u8).
+
+## Proposal
+
+Based on the investigation above, we propose to define the Dropout operation in
+the oneDNN Graph API as follows:
+
+**Operation Kind:**
+
+- `Dropout` (C++), `dnnl_graph_op_dropout` (C)
+
+**Inputs:**
+
+- `src`: input tensor (`f32` / `bf16` / `f16`)
+- `seed`: random seed (`uint64_t`, single value)
+- `offset`: random offset (`uint64_t`, single value)
+- `p`: dropout probability (`f32`, single value)
```

Comment:
Does that mean seed, offset, and p will be changed among different iterations?

Reply:
p and seed most likely will not change among different iterations, only offset will change

Reply:
Then why do we put p and seed as inputs, not attributes? Are they on device memory?

Reply:
Extra things to consider here wrt input vs attribute: using attribute will also trigger extra jitting and increase cache occupancy. So unless we jit completely different code based on p and seed (which we don't), it might be beneficial to keep those as inputs.

If those are never on device, would it make sense to only support host_side_scalar for those?